Protect Meets Core: add components to scan package

nateweller opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Pulls out the components from https://github.com/Automattic/jetpack/pull/40191 and adds them to the `@automattic/jetpack-scan` project.
* Includes some enhancements and fixes primarily for component storybooks.
* After this is merged, the `add/protect/core` branch can be rebased, and any modifications in the components package removed.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

peb6dq-31T-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Review the new "Scan" section in Storybook: `cd projects/js-packages/storybook && pnpm storybook:dev` to run locally.
* Smoke test the Jetpack Protect plugin, validate the threat severity badge renders appropriately on the client side.

<img width="827" alt="Screenshot 2025-03-19 at 12 48 20 PM" src="https://github.com/user-attachments/assets/0ecf91ed-bdb1-4d02-b5b8-fedca695b01a" />

<img width="1204" alt="Screenshot 2025-03-19 at 12 46 44 PM" src="https://github.com/user-attachments/assets/4ef13091-0282-4513-a5b4-1c9dd162d3e0" />

nateweller committed:
Move threat components from jetpack-components to jetpack-scan

changelog

changelog

Update changelog

Update changelog

changelog

nateweller committed:
Fix package versions

nateweller committed:
fix deps sort order

nateweller committed:
Fix tests

nateweller committed:
Migrate components from add/protect/core into the scan package

nateweller committed:
Component enhancements and fixes

nateweller committed:
changelog

dkmyta reviewed (commented):
Really nicely done! A few minor things:
- The **Loading**, **Stale**, and **Error** stories for the `ThreatFixerButton` do not appear to be working.
- For the `ThreatsDataViews` **Historic** story, I noticed we offer a `Stop ignoring` item in the action drop down, will this immediately trigger an un-ignore without confirmation or should it open an un-ignore confirmation modal?
- The `ThreatsDataViews` **Additional Connections Needed** story currently does not show any prompts for connections when triggering an action.
- The `ThreatsDataViews` **User Connection Needed** story currently shows the credentials prompt when triggering an action.
- I think this might be correct but I will mention it anyway, for all of the `ThreatsDataViews` connection related stories we don't show the connection prompts for ignore/unignore, my memory is failing me here, do we not currently do that or need that for these actions?
- The `ThreatsDataViews` **Free Results** story does not currently render the upgrade prompt in the modal.

Some non-blocking nit-picks: 
- For the `ThreatsDataViews`, is there any way to improve the transition between modal states? Looks like currently the modal closes, we quickly flash the table, then a new one opens with the new content, rather than modal remaining open and the content changing without a transition.
- For the `ThreatsDataViews` **Historic** story I initially found the `Stop ignoring` button content slightly confusing for the un-ignored threats modal - as if it was currently in the process of ignoring and I could stop that. I don't love `Un-ignore` either but it is the clear opposite action of `Ignore`.

nateweller committed:
Fix threat fixer button stories

nateweller committed:
Remove use of TS for tests

nateweller committed:
'Stop Ignoring' -> 'Un-ignore'

nateweller committed:
Use storybook argTypes for ThreatsDataViews storybook

nateweller commented:
@dkmyta I've updated the ThreatsDataViews story to use storybook args to allow for testing various component props together:

<img width="1039" alt="Screenshot 2025-03-29 at 12 23 58 PM" src="https://github.com/user-attachments/assets/e35c9fdf-0b61-4809-8b6b-7938510a494d" />

Additionally fixed the fixer button stories, and adjusted stop vs un ignore verbiage :+1:

dkmyta reviewed (commented):
`ThreatsDataViews` stories are much better using the storybook args, nice work! 

I haven't exhaustively tested all the various combinations but there seems to be at least one outstanding issue with this configuration:
```
Data Type : Current Threat
Jetpack Plan : Scan
Jetpack Connection : Custom
Site Credentials : Connected
Refer to Codeable : false
```

Clicking an update action causes `Error: Cannot read properties of null (reading 'connected')`

I am not sure of the use-case of this `Custom` connection type, and it might not be a possible combination in the wild, but we should ensure stories aren't erroring out in any case, and/or remove any options that aren't currently possible.

Previously we had a custom modal state for when there were both connections missing. Is that no longer a thing? I have no problem with the individual/consecutive connection modals, just want to make sure I am not missing something I should be seeing.

A few items from my previous notes we should also confirm on:
- Should `Un-ignore` (from modal or actions dropdown) immediately trigger an un-ignore action or should it open a un-ignore confirmation modal like the ignore action?
- Ignore actions prompt for connections, should un-ignore actions?
   - Note: Seems like only a site connection is prompted currently
   - Question: If only fixers require connections, maybe both ignore/un-ignore can be free of this? Either way, we should keep them consistent IMO
- I suspect improving modal transitions is not an easy fix, just wanted to get your opinions in case you missed it initially
